admittance_controller/include/admittance_controller/admittance_rule_impl.hpp

```diff
@@ -0,0 +1,444 @@
+  admittance_state.joint_pos += admittance_state.joint_vel * dt;
+
+  // calculate admittance velocity corresponding to joint velocity
+  success &= kinematics_->convert_joint_deltas_to_cartesian_deltas(
+    admittance_state.current_joint_pos, admittance_state.joint_vel,
+    admittance_state.ft_sensor_frame, admittance_state.admittance_velocity);
+  success &= kinematics_->convert_joint_deltas_to_cartesian_deltas(
+    admittance_state.current_joint_pos, admittance_state.joint_acc,
+    admittance_state.ft_sensor_frame, admittance_state.admittance_acceleration);
+
+  return success;
+}
+
+void AdmittanceRule::process_wrench_measurements(
+  const geometry_msgs::msg::Wrench & measured_wrench,
+  const Eigen::Matrix<double, 3, 3> & sensor_world_rot,
+  const Eigen::Matrix<double, 3, 3> & cog_world_rot)
+{
+  Eigen::Matrix<double, 3, 2, Eigen::ColMajor> new_wrench;
+  new_wrench(0, 0) = measured_wrench.force.x;
+  new_wrench(1, 0) = measured_wrench.force.y;
+  new_wrench(2, 0) = measured_wrench.force.z;
+  new_wrench(0, 1) = measured_wrench.torque.x;
+  new_wrench(1, 1) = measured_wrench.torque.y;
+  new_wrench(2, 1) = measured_wrench.torque.z;
+
+  // transform to world frame
+  Eigen::Matrix<double, 3, 2> new_wrench_base = sensor_world_rot * new_wrench;
+
+  // apply gravity compensation
+  new_wrench_base(2, 0) -= ee_weight_[2];
+  new_wrench_base.block<3, 1>(0, 1) -= (cog_world_rot * cog_).cross(ee_weight_);
+
+  // apply smoothing filter
+  for (size_t i = 0; i < 6; ++i)
+  {
+    wrench_world_(i) = filters::exponentialSmoothing(
+      new_wrench_base(i), wrench_world_(i), parameters_.ft_sensor.filter_coefficient);
+  }
+}
+
+const control_msgs::msg::AdmittanceControllerState & AdmittanceRule::get_controller_state()
+{
+  for (size_t i = 0; i < parameters_.joints.size(); ++i)
+  {
+    state_message_.joint_state.name[i] = parameters_.joints[i];
+  }
+  for (size_t i = 0; i < admittance_state_.joint_pos.size(); ++i)
+  {
+    state_message_.joint_state.position[i] = admittance_state_.joint_pos[i];
+  }
+  for (size_t i = 0; i < admittance_state_.joint_vel.size(); ++i)
+  {
+    state_message_.joint_state.velocity[i] = admittance_state_.joint_vel[i];
+  }
+  for (size_t i = 0; i < admittance_state_.joint_acc.size(); ++i)
+  {
+    state_message_.joint_state.effort[i] = admittance_state_.joint_acc[i];
+  }
+  for (size_t i = 0; i < admittance_state_.stiffness.size(); ++i)
+  {
+    state_message_.stiffness.data[i] = admittance_state_.stiffness[i];
+  }
+  for (size_t i = 0; i < admittance_state_.damping.size(); ++i)
+  {
+    state_message_.damping.data[i] = admittance_state_.damping[i];
+  }
+  for (size_t i = 0; i < admittance_state_.selected_axes.size(); ++i)
+  {
+    state_message_.selected_axes.data[i] = static_cast<bool>(admittance_state_.selected_axes[i]);
+  }
+  for (size_t i = 0; i < admittance_state_.mass.size(); ++i)
+  {
+    state_message_.mass.data[i] = admittance_state_.mass[i];
+  }
```

Comment:
This can be shorter, if not, you probably have some other problems, e.g., initialization

```suggestion
  for (size_t i = 0; i < parameters_.joints.size(); ++i)
  {
    state_message_.joint_state.name[i] = parameters_.joints[i];
    state_message_.joint_state.position[i] = admittance_state_.joint_pos[i];
    state_message_.joint_state.velocity[i] = admittance_state_.joint_vel[i];
    state_message_.joint_state.effort[i] = admittance_state_.joint_acc[i];
    state_message_.stiffness.data[i] = admittance_state_.stiffness[i];
    state_message_.damping.data[i] = admittance_state_.damping[i];
    state_message_.selected_axes.data[i] = static_cast<bool>(admittance_state_.selected_axes[i]);
    state_message_.mass.data[i] = admittance_state_.mass[i];
  }
```